tests/Utils/TestRunner.hs

```diff
@@ -56,16 +59,74 @@ data ApiResultCheck = ApiResultCheck
 
 runAll :: [Command T.Text] -> IO (HM.HashMap RequestKey ApiResult)
 runAll cmds = do
-  withAsync (serve _testConfigFilePath) $ \_ -> do
-    sendResp <- doSend $ SubmitBatch cmds
-    case sendResp of
-      ApiFailure _ -> return HM.empty
-      ApiSuccess RequestKeys{..} -> do
-        pollResp <- doPoll $ Poll _rkRequestKeys
-        case pollResp of
-          ApiFailure _ -> return HM.empty
-          ApiSuccess (PollResponses apiResults) -> do
-            return apiResults
+  result <- Exception.try helper 
+  case result of
+    Left (Exception.SomeException err) -> do
+      putStrLn "Exception occurred --- "
+      putStrLn (show err)
+      return HM.empty
+    Right res -> return res
+
+  where helper = Exception.bracket
+                 (startServer _testConfigFilePath)
+                 stopServer
+                 (const (run cmds))
+
+startServer :: FilePath -> IO (Async (), Async (), Async ())
+startServer configFile = do
+  flushDb
+  ((histC, inC, debugFn, port, logDir), (asyncCmd, asyncHist)) <- setupServer configFile
+  asyncServer <- async (runApiServer histC inC debugFn port logDir)
+  link2 asyncServer asyncCmd
+  link2 asyncServer asyncHist
+  return (asyncServer, asyncCmd, asyncHist)
+
+stopServer :: (Async (), Async (), Async ()) -> IO ()
+stopServer (asyncServer, asyncCmd, asyncHist) = do
+   print $ "canceling asyncCmd " ++ (show (asyncThreadId asyncCmd))
+   uninterruptibleCancel asyncCmd
+   cmdPoll <- poll asyncCmd
+   print $ "Result of cmdPoll --- " ++ show cmdPoll
+   
+   print $ "canceling asyncHist " ++ (show (asyncThreadId asyncHist))
+   uninterruptibleCancel asyncHist
+   histPoll <- poll asyncHist
+   print $ "Result of histPoll --- " ++ show histPoll
+          
+   print $ "canceling asyncServer " ++ (show (asyncThreadId asyncServer))
+   uninterruptibleCancel asyncServer
+   serverPoll <- poll asyncServer
+   print $ "Result of serverPoll --- " ++ show serverPoll
+
+   flushDb
+          
+
+run :: [Command T.Text] -> IO (HM.HashMap RequestKey ApiResult)
+run cmds = do
+  sendResp <- doSend $ SubmitBatch cmds
+  case sendResp of
+    ApiFailure _ -> do
+      print $ "API FAILURE on send/ --- " ++ (show . last) cmds
+      return HM.empty
+    ApiSuccess RequestKeys{..} -> do
+      results <- timeout 3000000 (helper _rkRequestKeys)
+      case results of
+        Nothing -> do
+          print $ "API SUCCESS with empty map. NOT Retrying --- " ++ (show . last) cmds
+          return HM.empty
+        Just res -> return res
+      
+  where helper reqKeys = do
+          pollResp <- doPoll $ Poll reqKeys
+          case pollResp of
+            ApiFailure _ -> do
+              print $ "API FAILURE on poll/" ++ (show . last) cmds
+              return HM.empty
+            ApiSuccess (PollResponses apiResults) -> do
+              if null apiResults then do
+                print $ "API SUCCESS with empty map. Retrying."
+                helper reqKeys
```

Comment:
Does recursive polling of results because sometimes the polling is done faster then the send results are written to database. This was one of the causes of CI errors.